File: clients/typescript/tests/packaging.test.ts
import { afterAll, beforeAll, describe, expect, test } from '@jest/globals';
import { execFileSync } from 'child_process';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';

/**
 * Drives the built package the way npm hands it to a consumer, through the `exports` map.
 *
 * `npm run check-full` type-checks and tests the sources; none of that says anything about whether
 * `require('hook0-client')` or `import 'hook0-client'` resolve to a file that exists. A package can
 * be perfectly written and still unusable — a condition pointing at a path the build does not emit,
 * a path the `files` list does not publish, an ES module half whose export list drifted from the
 * CommonJS one. Every one of those is invisible to a compiler and obvious to the first consumer who
 * installs it.
 *
 * So the package is installed here, near enough: a directory outside the tree with the package
 * linked under `node_modules`, which is all Node's resolver needs to read `exports` and pick a
 * condition exactly as it would after `npm install`. The probes below are separate processes
 * running real `require` and real `import`.
 */

const PACKAGE_ROOT = path.resolve(__dirname, '..');
const MANIFEST = JSON.parse(
  fs.readFileSync(path.join(PACKAGE_ROOT, 'package.json'), 'utf8')
) as Manifest;

/**
 * The same file read as the fields it carries rather than as the shape declared below, so that the
 * dependency guard looks at what appeared in the manifest instead of at what a type says may.
 */
const MANIFEST_FIELDS = JSON.parse(
  fs.readFileSync(path.join(PACKAGE_ROOT, 'package.json'), 'utf8')
) as Record<string, Record<string, string>>;

/**
 * The manifest fields npm installs into the tree of whoever depends on this package.
 * `devDependencies` is not one of them: it is never installed by a consumer.
 */
const INSTALLED_INTO_CONSUMERS = ['dependencies', 'peerDependencies', 'optionalDependencies'];

/** No probe here reaches the network or the disk beyond a handful of files; a slow one is a hung one. */
const PROBE_TIMEOUT_MS = 60_000;

/** Room for the probe to spawn on a loaded machine, without ever letting a hung one hold the suite. */
const TEST_TIMEOUT_MS = 120_000;

/** What this test reads out of `package.json`. */
interface Manifest {
  name: string;
  main: string;
  module: string;
  types: string;
  files: string[];
  exports: Record<string, ExportCondition | string>;
}

type ExportCondition = { [condition: string]: ExportCondition | string };

let consumer: string;

/**
 * A directory outside the package with the package linked under `node_modules`, so that a bare
 * `hook0-client` specifier resolves through the `exports` map rather than through a relative path.
 */
beforeAll(() => {
  consumer = fs.mkdtempSync(path.join(os.tmpdir(), 'hook0-client-packaging-'));
  fs.mkdirSync(path.join(consumer, 'node_modules'), { recursive: true });
  fs.symlinkSync(PACKAGE_ROOT, path.join(consumer, 'node_modules', MANIFEST.name), 'dir');
  fs.writeFileSync(
    path.join(consumer, 'package.json'),
    JSON.stringify({ name: 'packaging-probe', version: '0.0.0', private: true, type: 'module' })
  );
});

afterAll(() => {
  if (consumer !== undefined) {
    fs.rmSync(consumer, { recursive: true, force: true });
  }
});

/** Runs one probe in its own process and returns what it printed. */
function probe(fileName: string, source: string): string {
  fs.writeFileSync(path.join(consumer, fileName), source);
  return execFileSync(process.execPath, [fileName], {
    cwd: consumer,
    timeout: PROBE_TIMEOUT_MS,
    encoding: 'utf8',
    stdio: ['ignore', 'pipe', 'pipe'],
  }).trim();
}

/** Every path an `exports` map, `main`, `module` or `types` points at, flattened. */
function declaredPaths(): string[] {
  const found: string[] = [MANIFEST.main, MANIFEST.module, MANIFEST.types];

  const walk = (node: ExportCondition | string): void => {
    if (typeof node === 'string') {
      found.push(node);
      return;
    }
    Object.values(node).forEach(walk);
  };
  Object.values(MANIFEST.exports).forEach(walk);

  return [...new Set(found.map((candidate) => candidate.replace(/^\.\//, '')))].sort();
}

describe('published package', () => {
  test('installs nothing into a consumer', () => {
    // Every client in this repository reaches the network, signs and decodes with what its runtime
    // already ships, and that sentence is worth exactly what the guard behind it is worth. A
    // dependency here is installed by everyone who installs `hook0-client`, whether they wanted it
    // or not — including a bundler polyfill standing in for a builtin, which is how the `url`
    // package got here in the first place.
    const declared = Object.entries(MANIFEST_FIELDS)
      .filter(([field]) => INSTALLED_INTO_CONSUMERS.includes(field))
      .flatMap(([field, named]) => Object.keys(named).map((name) => `${field}.${name}`));

    expect({ declared }).toEqual({ declared: [] });
  });

  test('points every condition at a file the build produces', () => {
    const missing = declaredPaths().filter(
      (declared) => !fs.existsSync(path.join(PACKAGE_ROOT, declared))
    );

    expect({ missing }).toEqual({ missing: [] });
  });

  test('publishes every file a condition points at', () => {
    // `files` entries are anchored at the package root, with or without a leading slash; npm always
    // ships `package.json` whatever the list says.
    const published = new Set(MANIFEST.files.map((entry) => entry.replace(/^\//, '')));

    const unpublished = declaredPaths().filter(
      (declared) => declared !== 'package.json' && !published.has(declared.split('/')[0])
    );

    expect({ unpublished }).toEqual({ unpublished: [] });
  });

  test(
    'hands both module formats the same names, backed by the same objects',
    () => {
      const printed = probe(
        'both.mjs',
        `import { createRequire } from 'module';
         import * as esm from 'hook0-client';
         const require = createRequire(import.meta.url);
         const cjs = require('hook0-client');

         const names = Object.keys(cjs).sort();
         console.log(JSON.stringify({
           esm: Object.keys(esm).filter((name) => name !== 'default').sort(),
           cjs: names,
           differing: names.filter((name) => esm[name] !== cjs[name]),
           defaultIsWholeModule: esm.default === cjs,
           esmFile: import.meta.resolve('hook0-client'),
           cjsFile: require.resolve('hook0-client'),
         }));`
      );

      const resolved = JSON.parse(printed) as {
        esm: string[];
        cjs: string[];
        differing: string[];
        defaultIsWholeModule: boolean;
        esmFile: string;
        cjsFile: string;
      };

      // The names the ES module half declares are the names the CommonJS half has: neither list may
      // drift ahead of the other.
      expect(resolved.esm).toEqual(resolved.cjs);
      expect(resolved.esm.length).toBeGreaterThan(0);

      // And they are the same objects, not equal-looking copies of them. This is what keeps
      // `instanceof Hook0ClientError` answering true across a dependency tree that imported the
      // package both ways.
      expect({ differing: resolved.differing }).toEqual({ differing: [] });

      // `import hook0Client from 'hook0-client'` keeps meaning what it meant before the ES module
      // half existed: the whole module.
      expect(resolved.defaultIsWholeModule).toBe(true);

      // Each condition led somewhere different, or the map is not doing anything.
      expect(resolved.esmFile).toContain(MANIFEST.module);
      expect(resolved.cjsFile).toContain(path.normalize(MANIFEST.main));
      expect(resolved.esmFile).not.toContain(MANIFEST.main);
    },
    TEST_TIMEOUT_MS
  );

  test(
    'keeps the compiled layout out of the contract',
    () => {
      // With an `exports` map in place, the compiled tree stops being reachable — a consumer who
      // reached into `hook0-client/dist/lib` is told so at resolution rather than being quietly
      // handed a second copy of the module. That refusal is the map proving it is in force.
      const printed = probe(
        'deep.mjs',
        `import { createRequire } from 'module';
         const require = createRequire(import.meta.url);
         try {
           require('hook0-client/dist/lib');
           console.log(JSON.stringify({ code: 'RESOLVED' }));
         } catch (refusal) {
           console.log(JSON.stringify({ code: refusal.code }));
         }`
      );

      expect(JSON.parse(printed)).toEqual({ code: 'ERR_PACKAGE_PATH_NOT_EXPORTED' });
    },
    TEST_TIMEOUT_MS
  );
});

File: clients/zig/tests/packaging_test.zig
//! What this package claims to be, held against what it is.
//!
//! Zig has no central registry to reject a manifest that does not match its package, so the manifest
//! is checked here instead: that the version it releases is the version the code answers, that the
//! Zig it declares is the Zig the pipeline installs, that the files it ships are the files that are
//! here, and — the one worth the most — that it still depends on nothing.
//!
//! That last one is a promise a reader cannot verify by reading, since a dependency is one line in a
//! file nobody opens twice. Here it is a type: the manifest is parsed into a struct whose
//! `dependencies` has no fields, so a dependency appearing is a parse that fails.

const std = @import("std");
const builtin = @import("builtin");
const hook0 = @import("hook0");

const io = std.testing.io;
const allocator = std.testing.allocator;

/// Largest file this case reads back, in bytes.
const max_manifest_bytes: usize = 64 * 1024;

/// Where the pipeline that builds this package is defined, relative to the package root.
const pipeline_path = "../../.gitlab-ci.yml";

/// What names the Zig the pipeline installs, in that file.
const pipeline_zig_version = "ZIG_VERSION:";

/// What the build writes rather than what the package holds, and so what is not shipped.
///
/// Entries starting with a dot are left out too — the cache and the pipeline definition both are —
/// which is why this names only the one build output that does not.
const build_output = "zig-out";

/// `build.zig.zon`, as far as this case is concerned with it.
///
/// `dependencies` is deliberately a struct with no fields: `std.zon.parse` refuses a member it was
/// not declared to hold, so the day somebody adds one, this stops being a manifest this case can
/// read at all.
const Manifest = struct {
    name: enum { hook0_client },
    version: []const u8,
    minimum_zig_version: []const u8,
    fingerprint: u64,
    dependencies: struct {},
    paths: []const []const u8,
};

fn manifestOf(arena: std.mem.Allocator) !Manifest {
    const written = try std.Io.Dir.cwd().readFileAlloc(io, "build.zig.zon", arena, .limited(max_manifest_bytes));
    const terminated = try arena.dupeZ(u8, written);

    var diagnostics: std.zon.parse.Diagnostics = .{};
    return std.zon.parse.fromSliceAlloc(Manifest, arena, terminated, &diagnostics, .{}) catch |failed| {
        std.debug.print("build.zig.zon does not say what this package declares: {f}\n", .{diagnostics});
        return failed;
    };
}

test "the manifest declares no dependencies at all" {
    var arena: std.heap.ArenaAllocator = .init(allocator);
    defer arena.deinit();

    // Reading it is the assertion: a manifest carrying a dependency is one `Manifest` cannot hold.
    const manifest = try manifestOf(arena.allocator());
    try std.testing.expectEqual(@as(usize, 0), @typeInfo(@TypeOf(manifest.dependencies)).@"struct".fields.len);
}

test "the manifest releases the version this package answers" {
    var arena: std.heap.ArenaAllocator = .init(allocator);
    defer arena.deinit();

    const manifest = try manifestOf(arena.allocator());
    try std.testing.expectEqualStrings(hook0.version, manifest.version);
}

test "the Zig the manifest declares is the Zig the pipeline installs, and the one running this" {
    var arena: std.heap.ArenaAllocator = .init(allocator);
    defer arena.deinit();
    const held = arena.allocator();

    const manifest = try manifestOf(held);
    const declared = try std.SemanticVersion.parse(manifest.minimum_zig_version);

    const written = try std.Io.Dir.cwd().readFileAlloc(io, pipeline_path, held, .limited(max_manifest_bytes));
    var lines = std.mem.splitScalar(u8, written, '\n');
    var pinned: ?[]const u8 = null;
    while (lines.next()) |line| {
        const at = std.mem.indexOf(u8, line, pipeline_zig_version) orelse continue;
        pinned = std.mem.trim(u8, line[at + pipeline_zig_version.len ..], " \"'\r\t");
    }

    std.testing.expect(pinned != null) catch |failed| {
        std.debug.print("{s} names no {s}\n", .{ pipeline_path, pipeline_zig_version });
        return failed;
    };
    try std.testing.expectEqualStrings(manifest.minimum_zig_version, pinned.?);

    // The compiler running this case, held to the same two numbers. A patch release is a compiler
    // somebody may reasonably be on; a different minor is the one where `std.Io` changed shape,
    // which is what this package is written against.
    try std.testing.expectEqual(declared.major, builtin.zig_version.major);
    try std.testing.expectEqual(declared.minor, builtin.zig_version.minor);
}

test "the manifest ships every file of this package, and nothing that is not here" {
    var arena: std.heap.ArenaAllocator = .init(allocator);
    defer arena.deinit();
    const held = arena.allocator();

    const manifest = try manifestOf(held);

    // What is here, found by looking rather than by naming: a directory somebody adds joins this
    // case without anybody remembering to add it.
    var root = try std.Io.Dir.cwd().openDir(io, ".", .{ .iterate = true });
    defer root.close(io);

    var present: std.ArrayList([]const u8) = .empty;
    var walked = root.iterateAssumeFirstIteration();
    while (try walked.next(io)) |entry| {
        if (std.mem.startsWith(u8, entry.name, ".")) continue;
        if (std.mem.eql(u8, entry.name, build_output)) continue;
        try present.append(held, try held.dupe(u8, entry.name));
    }

    for (present.items) |name| {
        var shipped = false;
        for (manifest.paths) |path| shipped = shipped or std.mem.eql(u8, path, name);
        std.testing.expect(shipped) catch |failed| {
            std.debug.print("`{s}` is in this package and `build.zig.zon` does not ship it\n", .{name});
            return failed;
        };
    }

    for (manifest.paths) |path| {
        var here = false;
        for (present.items) |name| here = here or std.mem.eql(u8, path, name);
        std.testing.expect(here) catch |failed| {
            std.debug.print("`build.zig.zon` ships `{s}`, which is not in this package\n", .{path});
            return failed;
        };
    }
}
